tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1021 @@
+    }
+  }
+
+  /**
+   * Initialize truePositives, falsePositives, trueNegatives, and falseNegatives variables, given
+   * the shape of the data.
+   *
+   * @param shape the prediction shape if called from updateState, otherwise null
+   */
+  @SuppressWarnings("unchecked")
+  private Map<ConfusionMatrixEnum, Assign<T>> build(Shape shape) {
+    Shape variableShape;
+    if (initialized) {
+      return Collections.EMPTY_MAP;
+    }
+
+    if (this.isMultiLabel()) {
+      if (shape == null) {
+        throw new IllegalArgumentException("For multiLabel, a shape must be provided");
+      }
+      if (shape.numDimensions() != 2)
+        throw new IllegalArgumentException(
+            String.format(
+                "labels must have rank=2 when multiLabel is true. Found rank %d.",
+                shape.numDimensions()));
+      this.numLabels = (int) shape.size(1);
+      variableShape = Shape.of(this.numThresholds, this.numLabels);
+    } else {
+      variableShape = Shape.of(this.numThresholds);
+    }
+
+    Zeros<T> zeros = new Zeros<>(getTF());
+    Operand<T> zero = zeros.call(getTF().constant(variableShape), type);
+    if (truePositives == null) {
+      truePositives = getTF().withName(getTruePositivesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.TRUE_POSITIVES, getTF().assign(truePositives, zero));
+    }
+
+    if (falsePositives == null) {
+      falsePositives = getTF().withName(getFalsePositivesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.FALSE_POSITIVES, getTF().assign(falsePositives, zero));
+    }
+
+    if (trueNegatives == null) {
+      trueNegatives = getTF().withName(getTrueNegativesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.TRUE_NEGATIVES, getTF().assign(trueNegatives, zero));
+    }
+
+    if (falseNegatives == null) {
+      falseNegatives = getTF().withName(getFalseNegativesName()).variable(zero);
+      initializers.put(ConfusionMatrixEnum.FALSE_NEGATIVES, getTF().assign(falseNegatives, zero));
+    }
+
+    this.initialized = true;
+    return initializers;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @SuppressWarnings("unchecked")
+  public List<Op> updateStateList(
+      Operand<? extends TNumber> labels,
+      Operand<? extends TNumber> predictions,
+      Operand<? extends TNumber> sampleWeights) {
+
+    Operand<T> lLabels = cast(getTF(), labels, type);
+    Operand<T> lPredictions = cast(getTF(), predictions, type);
+    Operand<T> tSampleWeights = sampleWeights != null ? cast(getTF(), sampleWeights, type) : null;
+    List<Op> updateOperations = new ArrayList<>();
+    Map<ConfusionMatrixEnum, Assign<T>> varInitializers = Collections.EMPTY_MAP;
+    if (!this.initialized) {
+      varInitializers = build(lPredictions.shape());
+    }
+    if (this.isMultiLabel() || this.getLabelWeights() != null) {
+      List<SymbolicShape<? extends TNumber>> symbols = new ArrayList<>();
+      symbols.add(new SymbolicShape<>(lLabels, "N", "L"));
```

Comment:
Symbolic shapes was a challenge to figure out. What it boils down to is a set of rules to later on verify that the shapes at run time meet a set of rules. If the condition  is `multiLabel` and has `labelWeights`, the shape of the labels must be `Number of Examples ("N")` and `Number of Labels ("L")`.
Each variable then must be `Number of Thresholds ("T")` and `Number of Labels ("L")`. The shape of `labelWeights`, if not null, must be `Number of Labels ("L")`. `assertShapes() `adds a collection of `assertThat` operations as control Ops to do the realtime checking by substituting the real time values for "L", "T" and "N" to make sure all the Operands are consistent with these rules. 

The rules get applied to labels, and the variables and labelWeights. 

Later on in `updateConfusionMatrixVariables()`, the rules would be asserted through control dependencies. But `updateConfusionMatrixVariables()` knows nothing of that, so I am not sure how to avoid handling the squeeze/compress. 

I think more code notes would be helpful, but TF Python is not very verbose, so I am just trying to mimic their code as best I can. I can only assume they had a reason for writing the code they way they did. Also, there is a necessary translation from Pyhton-ese to Java, that is not always straight forward. With that said, the more we can figure out what they intended to do the better.

Reply:
I have changed `updateConfusionMatrixVariables` `thresholds` argument from` float[]` to `Operand<TFloat32>`.

Reply:
👍 done